File: 4.Authentication and authorization for backend engineers/code/typescript/05_json_web_tokens_jwt_4.ts
import jwt from 'jsonwebtoken';
import { Redis } from 'ioredis';
import crypto from 'crypto';

// Verify with a PUBLIC key, pinning RS256 (blocks alg:none & key confusion)
export function verifyRS256(tokenStr: string, pub: string | Buffer): any {
    try {
        return jwt.verify(tokenStr, pub, { algorithms: ['RS256'] });
    } catch (err) {
        throw new Error('Invalid signature or expired token');
    }
}

function newSessionId(): string {
    return crypto.randomBytes(32).toString('hex');
}

// Rotate a refresh token; detect reuse of an already-spent one.
export async function refresh(rdb: Redis, presented: string): Promise<string> {
    const family = await rdb.get(`rt:${presented}`);
    if (!family) {
        // not a live token ,  was it a previously-spent one? -> theft
        const fam = await rdb.get(`spent:${presented}`);
        if (fam) {
            await rdb.del(`family:${fam}`); // revoke the whole family
            throw new Error("refresh reuse detected");
        }
        throw new Error("invalid refresh token");
    }
    
    await rdb.del(`rt:${presented}`); // consume the live token
    await rdb.set(`spent:${presented}`, family, 'EX', 14 * 24 * 60 * 60); // remember it
    
    const newRefresh = newSessionId();
    await rdb.set(`rt:${newRefresh}`, family, 'EX', 14 * 24 * 60 * 60);
    return newRefresh;
}

File: 4.Authentication and authorization for backend engineers/code/rust/05_json_web_tokens_jwt_4.rs
use jsonwebtoken::{decode, Validation, DecodingKey, errors::Error};
use serde::{Deserialize, Serialize};
use redis::{AsyncCommands, Client};
use rand::RngCore;

#[derive(Debug, Serialize, Deserialize)]
pub struct Claims {
    // Define claims as needed
    pub sub: String,
}

// Verify with a PUBLIC key, pinning RS256 (blocks alg:none & key confusion)
pub fn verify_rs256(token_str: &str, pub_key: &[u8]) -> Result<Claims, Error> {
    let mut validation = Validation::new(jsonwebtoken::Algorithm::RS256);
    let token_data = decode::<Claims>(
        token_str,
        &DecodingKey::from_rsa_pem(pub_key)?,
        &validation,
    )?;
    Ok(token_data.claims)
}

pub fn new_session_id() -> String {
    let mut b = [0u8; 32];
    rand::thread_rng().fill_bytes(&mut b); // cryptographically random
    hex::encode(b)
}

// Rotate a refresh token; detect reuse of an already-spent one.
pub async fn refresh(rdb: &Client, presented: &str) -> Result<String, Box<dyn std::error::Error>> {
    let mut con = rdb.get_async_connection().await?;
    
    let family: Option<String> = con.get(format!("rt:{}", presented)).await?;
    
    let family = match family {
        Some(f) => f,
        None => {
            // not a live token ,  was it a previously-spent one? -> theft
            let spent: Option<String> = con.get(format!("spent:{}", presented)).await?;
            if let Some(fam) = spent {
                let _: () = con.del(format!("family:{}", fam)).await?; // revoke the whole family
                return Err("refresh reuse detected".into());
            }
            return Err("invalid refresh token".into());
        }
    };
    
    let _: () = con.del(format!("rt:{}", presented)).await?; // consume the live token
    let _: () = con.set_ex(format!("spent:{}", presented), &family, 14 * 24 * 60 * 60).await?; // remember it
    
    let new_refresh = new_session_id();
    let _: () = con.set_ex(format!("rt:{}", new_refresh), &family, 14 * 24 * 60 * 60).await?;
    
    Ok(new_refresh)
}
